subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c

```diff
@@ -530,19 +541,79 @@ static int cell_pos_cb_send(const char *const rx_buf)
 	return -EFTYPE;
 }
 
+void agps_process(const char * const buf, const size_t buf_len)
+{
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+	int ret = nrf_cloud_agps_process(buf, buf_len);
+
+	LOG_DBG("A-GPS data processed; result: %d", ret);
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	if (ret == 0) {
+		ret = nrf_cloud_pgps_notify_prediction();
+		if (ret) {
+			LOG_ERR("Error requesting P-GPS notification: %d", ret);
+		}
+	}
+#endif
+#endif
+}
+
+void pgps_process(const char * const buf, const size_t buf_len)
+{
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	int ret = nrf_cloud_pgps_process(buf, buf_len);
+
+	LOG_DBG("P-GPS data processed; result: %d", ret);
+#endif
+}
+
+int cell_pos_process(const char * const buf)
+{
+#if defined(CONFIG_NRF_CLOUD_CELL_POS)
+	if (cell_pos_cb_send(buf) == 0) {
+		LOG_DBG("Cellular positioning data sent to provided callback");
+		return 0;
+	}
+#endif
+	return -ENOSYS;
+}
+
 static int dc_rx_data_handler(const struct nct_evt *nct_evt)
 {
+	__ASSERT_NO_MSG(nct_evt != NULL);
+	__ASSERT_NO_MSG(nct_evt->param.dc != NULL);
+
+	bool discon_req = false;
+
 	struct nrf_cloud_evt cloud_evt = {
-		.type = NRF_CLOUD_EVT_RX_DATA,
 		.data = nct_evt->param.dc->data,
 		.topic = nct_evt->param.dc->topic,
 	};
 
-	bool discon_req = nrf_cloud_detect_disconnection_request(nct_evt->param.dc->data.ptr);
-
-	/* All data is forwared to the app... unless a callback is registered */
-	if (cell_pos_cb_send(nct_evt->param.dc->data.ptr) == 0) {
+	switch (nrf_cloud_decode_dc_rx_topic(cloud_evt.topic.ptr)) {
+	case NRF_CLOUD_RCV_TOPIC_AGPS:
+		agps_process(cloud_evt.data.ptr, cloud_evt.data.len);
 		return 0;
+	case NRF_CLOUD_RCV_TOPIC_PGPS:
+		pgps_process(cloud_evt.data.ptr, cloud_evt.data.len);
+		return 0;
+	case NRF_CLOUD_RCV_TOPIC_CELL_POS:
+		if (cell_pos_process(cloud_evt.data.ptr) == 0) {
+			/* Data was sent to cell pos cb, do not send to application */
+			return 0;
+		}
+		cloud_evt.type = NRF_CLOUD_EVT_RX_DATA_CELL_POS;
+		break;
+	case NRF_CLOUD_RCV_TOPIC_UNHANDLED:
+		LOG_DBG("Received data on unhandled topic: %s",
+			(char *)(cloud_evt.topic.ptr ? cloud_evt.topic.ptr : "NULL"));
+		/* Intentional fall-through */
+	case NRF_CLOUD_RCV_TOPIC_GENERAL:
+	default:
+		cloud_evt.type = NRF_CLOUD_EVT_RX_DATA_GENERAL;
+		discon_req = nrf_cloud_detect_disconnection_request(cloud_evt.data.ptr);
```

Comment:
I feel that this should have its own topic (a device control topic), but given that that would require further BE changes, maybe that would be for a round-two PR